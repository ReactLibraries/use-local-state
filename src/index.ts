/* eslint-disable @typescript-eslint/no-explicit-any */
import { useCallback, useEffect, useRef, useState } from 'react';

/**
 * Type for State control
 *
 * @export
 * @interface LocalState
 * @template T The type of value to use for state
 */
export interface LocalState<T> {
  dispatches: React.Dispatch<T>[];
  value: T;
}

/**
 * Create a state
 *
 * @export
 * @template T The type of value to use for state
 * @param {(T | (() => T))} value Initial value
 * @return Instances of state
 */
export const useLocalStateCreate: {
  <T>(value: T | (() => T)): LocalState<T>;
  <T = undefined>(value?: T): LocalState<T>;
} = <T>(value: T | (() => T)) => {
  return useRef<LocalState<T>>({
    dispatches: [],
    value: typeof value === 'function' ? (<() => T>value)() : value,
  }).current;
};
export const useCreateLocalState = useLocalStateCreate;

/**
 * Perform the same action as useState
 *
 * @export
 * @template T The type of value to use for state
 * @param {LocalState<T>} state The type of value to use for state
 * @return [value,dispatch]
 */
export const useLocalState = <T = undefined>(
  state: LocalState<T>
): [T, (value: T | ((value: T) => T)) => void] => {
  const [, dispatch] = useState<T>();
  useEffect(() => {
    state.dispatches = [...state.dispatches, dispatch];
    return () => {
      state.dispatches = state.dispatches.filter((d) => d !== dispatch);
    };
  }, [state]);
  const setState = useCallback(
    (value: T | ((value: T) => T)) => mutateLocalState(state, value),
    [state]
  );
  return [state.value, setState];
};

/**
 * Select and retrieve the value of state
 *
 * @export
 * @template T The type of value to use for state
 * @template K Type of the selected value
 * @param {LocalState<T>} state The type of value to use for state
 * @param {(value: T) => K} callback callbak to select the target state.
 * @return {*}  {K} Selected state
 */
export const useLocalSelector = <T, K>(state: LocalState<T>, callback: (value: T) => K): K => {
  const [value, setValue] = useState(() => callback(state.value));
  const dispatch = useCallback((value: T) => setValue(callback(value)), []);
  useEffect(() => {
    state.dispatches = [...state.dispatches, dispatch];
    return () => {
      state.dispatches = state.dispatches.filter((d) => d !== dispatch);
    };
  }, [state]);
  return value;
};

/**
 * Write a value to state
 *
 * @export
 * @template T The type of value to use for state
 * @param {LocalState<T>} state The type of value to use for state
 * @param {(T | ((value: T) => T))} value A value to set for state or a callback to set
 */
export const mutateLocalState = <T>(state: LocalState<T>, value: T | ((value: T) => T)) => {
  state.value = typeof value === 'function' ? (<(value: T) => T>value)(state.value) : value;
  state.dispatches.forEach((dispatch) => dispatch(state.value));
};

/**
 * Reducer to manipulate the state.
 *
 * @export
 * @template T The type of value to use for state
 * @template R Reducer
 * @template K Action
 * @param {LocalState<T>} state
 * @param {R} reducer
 * @return {*} dispatch
 */
export const useLocallReducer = <
  T,
  R extends (state: T, action: any) => T,
  K extends Parameters<R>[1]
>(
  state: LocalState<T>,
  reducer: R
) => {
  return useCallback(
    (action: K) => mutateLocalState(state, reducer(state.value, action)),
    [state, reducer]
  );
};
