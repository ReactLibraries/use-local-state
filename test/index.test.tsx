import React, { useEffect } from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import {
  useLocalState,
  useCreateLocalState,
  LocalState,
  useLocalSelector,
  useLocallReducer,
} from '../src/index';
import { mutateLocalState } from './../src/index';

let container: HTMLElement;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
});

it('initData-undefined', () => {
  const Component01 = ({ localState }: { localState: LocalState<undefined> }) => {
    const [value] = useLocalState(localState);
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = ({ localState }: { localState: LocalState<undefined> }) => {
    const [value] = useLocalState(localState);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = ({ localState }: { localState: LocalState<undefined> }) => {
    const [value] = useLocalState(localState);
    return <>{value ?? 'undefined'}</>;
  };
  const Parent = () => {
    const localState = useCreateLocalState();
    return (
      <>
        <Component01 localState={localState} />
        <Component02 localState={localState} />
        <Component03 localState={localState} />
      </>
    );
  };
  act(() => {
    render(<Parent />, container);
  });
  expect(container.childNodes).toMatchSnapshot();
});

it('initData-normal', () => {
  const Component01 = ({ localState }: { localState: LocalState<number> }) => {
    const [value] = useLocalState(localState);
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = ({ localState }: { localState: LocalState<number> }) => {
    const [value] = useLocalState(localState);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = ({ localState }: { localState: LocalState<number> }) => {
    const [value] = useLocalState(localState);
    return <>{value ?? 'undefined'}</>;
  };
  const Parent = () => {
    const localState = useCreateLocalState(100);
    return (
      <>
        <Component01 localState={localState} />
        <Component02 localState={localState} />
        <Component03 localState={localState} />
      </>
    );
  };
  act(() => {
    render(<Parent />, container);
  });

  expect(container.childNodes).toMatchSnapshot();
});

it('useEffect', () => {
  const Component01 = ({ localState }: { localState: LocalState<number | string> }) => {
    const [value] = useLocalState(localState);
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = ({ localState }: { localState: LocalState<number | string> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue(3);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = ({ localState }: { localState: LocalState<number | string> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue('Effect');
    }, []);
    return <>{value ?? 'undefined'}</>;
  };

  const Parent = () => {
    const localState = useCreateLocalState<number | string>(100);
    return (
      <>
        <Component01 localState={localState} />
        <Component02 localState={localState} />
        <Component03 localState={localState} />
      </>
    );
  };
  act(() => {
    render(<Parent />, container);
  });

  expect(container.childNodes).toMatchSnapshot();
});

it('add', () => {
  const Component01 = ({ localState }: { localState: LocalState<number> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue((v) => v + 1);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = ({ localState }: { localState: LocalState<number> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue((v) => v + 1);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = ({ localState }: { localState: LocalState<number> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue((v) => v);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };

  const Parent = () => {
    const localState = useCreateLocalState<number>(100);
    return (
      <>
        <Component01 localState={localState} />
        <Component02 localState={localState} />
        <Component03 localState={localState} />
      </>
    );
  };
  act(() => {
    render(<Parent />, container);
  });

  expect(container.childNodes).toMatchSnapshot();
});

it('selector', () => {
  type LocalStateType = { value1: number; value2: number };
  const Component01 = ({ localState }: { localState: LocalState<LocalStateType> }) => {
    const value = useLocalSelector(localState, (v) => v.value1);
    useEffect(() => {
      mutateLocalState(localState, (v) => ({ ...v, value1: v.value1 + 1 }));
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = ({ localState }: { localState: LocalState<LocalStateType> }) => {
    const value = useLocalSelector(localState, (v) => v.value2);
    useEffect(() => {
      mutateLocalState(localState, { ...localState.value, value2: localState.value.value2 + 1 });
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = ({ localState }: { localState: LocalState<LocalStateType> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue((v) => v);
    }, []);
    return <>{value.value1 + value.value2}</>;
  };

  const Parent = () => {
    const localState = useCreateLocalState<LocalStateType>({ value1: 10, value2: 20 });
    return (
      <>
        <Component01 localState={localState} />
        <Component02 localState={localState} />
        <Component03 localState={localState} />
      </>
    );
  };
  act(() => {
    render(<Parent />, container);
  });

  expect(container.childNodes).toMatchSnapshot();
});

it('init-function', () => {
  const Component01 = ({ localState }: { localState: LocalState<number> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue((v) => v + 1);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = ({ localState }: { localState: LocalState<number> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue((v) => v + 1);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = ({ localState }: { localState: LocalState<number> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue((v) => v);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };

  const Parent = () => {
    const localState = useCreateLocalState<number>(() => 100);
    return (
      <>
        <Component01 localState={localState} />
        <Component02 localState={localState} />
        <Component03 localState={localState} />
      </>
    );
  };
  act(() => {
    render(<Parent />, container);
  });

  expect(container.childNodes).toMatchSnapshot();
});

it('reducer', () => {
  const reducer = (state: number, action: { type: 'increment' | 'dbl-increment' }) => {
    switch (action.type) {
      case 'increment':
        return state + 1;
      case 'dbl-increment':
        return state + 2;
    }
  };

  const Component01 = ({ localState }: { localState: LocalState<number> }) => {
    const dispatch = useLocallReducer(localState, reducer);
    const [value] = useLocalState(localState);
    useEffect(() => dispatch({ type: 'increment' }), []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = ({ localState }: { localState: LocalState<number> }) => {
    const dispatch = useLocallReducer(localState, reducer);
    const [value] = useLocalState(localState);
    useEffect(() => dispatch({ type: 'dbl-increment' }), []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = ({ localState }: { localState: LocalState<number> }) => {
    const [value, setValue] = useLocalState(localState);
    useEffect(() => {
      setValue((v) => v);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };

  const Parent = () => {
    const localState = useCreateLocalState<number>(100);
    return (
      <>
        <Component01 localState={localState} />
        <Component02 localState={localState} />
        <Component03 localState={localState} />
      </>
    );
  };
  act(() => {
    render(<Parent />, container);
  });
  expect(container.childNodes).toMatchSnapshot();
});
