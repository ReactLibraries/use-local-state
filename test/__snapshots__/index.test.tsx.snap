// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add 1`] = `
NodeList [
  102,
  102,
  102,
]
`;

exports[`init-function 1`] = `
NodeList [
  102,
  102,
  102,
]
`;

exports[`initData-normal 1`] = `
NodeList [
  100,
  100,
  100,
]
`;

exports[`initData-undefined 1`] = `
NodeList [
  undefined,
  undefined,
  undefined,
]
`;

exports[`reducer 1`] = `
NodeList [
  103,
  103,
  103,
]
`;

exports[`selector 1`] = `
NodeList [
  11,
  21,
  32,
]
`;

exports[`useEffect 1`] = `
NodeList [
  Effect,
  Effect,
  Effect,
]
`;
